aptos-move/framework/aptos-framework/sources/permissioned_signer.move

```diff
@@ -0,0 +1,423 @@
+        if(!is_permissioned_signer(permissioned)) {
+            // Master signer has no permissions associated with it.
+            return
+        };
+        let addr = signer::address_of(&permission_signer(permissioned));
+        if(!exists<PermStorage>(addr)) {
+            return
+        };
+        simple_map::remove(
+            &mut borrow_global_mut<PermStorage>(addr).perms,
+            &copyable_any::pack(perm)
+        );
+    }
+
+    /// Another flavor of api to extract and store permissions
+    public fun extract_permission<PermKey: copy + drop + store>(
+        s: &signer,
+        weight: u256,
+        perm: PermKey
+    ): Permission<PermKey> acquires PermStorage {
+        assert!(check_permission_consume(s, weight, perm), error::permission_denied(ECANNOT_EXTRACT_PERMISSION));
+        Permission {
+            owner_address: signer::address_of(s),
+            key: perm,
+            capacity: weight,
+        }
+    }
+
+    public fun get_key<PermKey>(perm: &Permission<PermKey>): &PermKey {
+        &perm.key
+    }
+
+    public fun address_of<PermKey>(perm: &Permission<PermKey>): address {
+        perm.owner_address
+    }
+
+    public fun consume_permission<PermKey: copy + drop + store>(
+        perm: &mut Permission<PermKey>,
+        weight: u256,
+        perm_key: PermKey
+    ): bool {
+        if(perm.key != perm_key) {
+            return false
+        };
+        if(perm.capacity >= weight) {
+            perm.capacity = perm.capacity - weight;
+            return true
+        } else {
+            return false
+        }
+    }
+
+    public fun store_permission<PermKey: copy + drop + store>(
+        s: &signer,
+        perm: Permission<PermKey>
+    ) acquires PermStorage {
+        assert!(is_permissioned_signer(s), error::permission_denied(ENOT_PERMISSIONED_SIGNER));
+        let Permission { key, capacity, owner_address } = perm;
+
+        assert!(signer::address_of(s) == owner_address, error::permission_denied(E_PERMISSION_MISMATCH));
+
+        let permission_signer = permission_signer(s);
+        let permission_signer_addr = signer::address_of(&permission_signer);
+        if(!exists<PermStorage>(permission_signer_addr)) {
+            move_to(&permission_signer, PermStorage::V1 { perms: simple_map::new()});
+        };
+        let perms = &mut borrow_global_mut<PermStorage>(permission_signer_addr).perms;
+        let key = copyable_any::pack(key);
+        if(simple_map::contains_key(perms, &key)) {
+            let entry = simple_map::borrow_mut(perms, &key);
+            entry.capacity = entry.capacity + capacity;
+        } else {
+            simple_map::add(perms, key, StoredPermission::V1 { capacity })
+        }
+    }
+
+    // =====================================================================================================
+    // Native Functions
+
+    /// Creates a permissioned signer from an existing universal signer. The function aborts if the
```

Comment:
order of comments here is interleaved . seems like signer_from_permissioned_impl should be the first native function defined.